### PR TITLE
feat: redirect TS template to r-n

### LINKED ISSUE #1740: Redirect react-native-template-typescript to react-native
## Describe the Feature

From RN 0.71 RN will be the default template language, which removes the need for the `react-native-template-typescript` template. To ease the migration, as this way of starting a TS project is referenced all over the docs/tutorials/guides/etc, we could detect users using this template and treat it as `react-native`. Outputting extra warning about what's happening under the hood, so that users are aware.

Schedule for 11.0 release

### PR DESCRIPTION
**Pull request by adamTrz**
Summary:
---------

Fixes #1740 

Test Plan:
----------

Fork the branch and try to `init` new project with `react-native-template-typescript`. 
It should print the warning and default to normal r-n template.

### CODE REVIEW

File: packages/cli/src/commands/init/init.ts [hunk 1]
```diff
@@ -175,13 +176,31 @@ async function installDependencies({
   loader.succeed();
 }
 
+function checkVersionAndTemplate(options: Options, version: string): string {
+  const isTypescriptTemplate = options.template?.includes(
```
Comment (thymikee) [inline packages/cli/src/commands/init/init.ts:180]: Make it strict, not includes. We don't want to target other templates by accident

File: packages/cli/src/commands/init/init.ts [hunk 2]
```diff
@@ -175,13 +176,31 @@ async function installDependencies({
...
+    'react-native-template-typescript',
+  );
+
+  if (isTypescriptTemplate && options.template?.includes('@')) {
+    const typescriptTemplateVersion = options.template.split('@')[1];
+    if (semver.gte(typescriptTemplateVersion, '0.71.0')) {
+      logger.warn(
+        "Starting from verision '0.71' typescript is used by default.\nDiregarding template and using default react-native",
```
Comment (p-syche) [inline packages/cli/src/commands/init/init.ts:188]: You're missing an "S" in "Disregarding" :) 

But honestly for me, the first sentence alone is enough.

Comment (thymikee) [inline packages/cli/src/commands/init/init.ts:188]: There's also a typo in "verision". My suggestion:
```suggestion
        "Ignoring custom template: 'react-native-template-typescript'. Starting from React Native v0.71 TypeScript is used by default.",
```

File: packages/cli/src/commands/init/init.ts [hunk 3]
```diff
@@ -175,13 +176,31 @@ async function installDependencies({
...
+}
+
 async function createProject(
   projectName: string,
   directory: string,
   version: string,
   options: Options,
 ) {
-  const templateUri = options.template || `react-native@${version}`;
+  const templateUri = checkVersionAndTemplate(options, version);
```
Comment (thymikee) [inline packages/cli/src/commands/init/init.ts:203]: let's rename this fn to `createTemplateUri`

**Review by MateWW — APPROVED** [review]
I've tested and it works perfectly!
Thanks! 🚀🥳
Screenshot 2022-11-07 at 14 20 59

File: packages/cli/src/commands/init/init.ts [hunk 4]
```diff
@@ -175,13 +176,31 @@ async function installDependencies({
...
 }
 
+function createTemplateUri(options: Options, version: string): string {
+  const isTypescriptTemplate = options.template?.includes(
+    'react-native-template-typescript',
+  );
+
+  if (isTypescriptTemplate && options.template?.includes('@')) {
+    const typescriptTemplateVersion = options.template.split('@')[1];
+    if (semver.gte(typescriptTemplateVersion, '0.71.0')) {
```
Comment (radko93) [inline packages/cli/src/commands/init/init.ts:186]: Not sure if I get this right. Is it the RN version inside the template? The TS template does not follow RN versioning. You can look up what are the versions here https://github.com/react-native-community/react-native-template-typescript#react-native--template-version

If I'm wrong just ignore this comment.

Comment (adamTrz) [reply packages/cli/src/commands/init/init.ts:186]: > Not sure if I get this right. Is it the RN version inside the template? The TS template does not follow RN versioning. You can look up what are the versions here https://github.com/react-native-community/react-native-template-typescript#react-native--template-version
> 
> If I'm wrong just ignore this comment.

No, you're 💯 correct.
Actually now I'm tending to just doing the check if there's TS template passed but without any version
If users go with `init --template react-native-template-typescript` once r-n 0.71 is released we will be using `react-native`. But if they pass version with template I would assume that they want lower react native version but with TS so we will allow that? 
wdyt?

Comment (radko93) [reply packages/cli/src/commands/init/init.ts:186]: Yeah, looks like that's the way to go!

**Review by radko93 — APPROVED** [review]

**Comment by TMisiukiewicz** [comment]
image

tested the expected behavior, works correctly ✅

**Review by TMisiukiewicz — APPROVED** [review]

**Review by TMisiukiewicz — APPROVED** [review]

**Pull request merged**